toolchain/check/convert.cpp

```diff
@@ -828,29 +828,21 @@ static auto CanRemoveQualifiers(SemIR::TypeQualifiers quals,
 static auto DiagnoseConversionFailureToConstraintValue(
     Context& context, SemIR::LocId loc_id, SemIR::InstId expr_id,
     SemIR::TypeId target_type_id) -> void {
-  CARBON_DCHECK(target_type_id == SemIR::TypeType::TypeId ||
-                context.types().Is<SemIR::FacetType>(target_type_id));
-
-  auto type_of_expr_id = context.insts().Get(expr_id).type_id();
-  CARBON_CHECK(context.types().IsFacetType(type_of_expr_id));
-  // If the source type is/has a facet value, then we can include its
-  // FacetType in the diagnostic to help explain what interfaces the
-  // source type implements.
-  auto facet_value_inst_id = SemIR::InstId::None;
-  if (auto facet_access_type =
-          context.insts().TryGetAs<SemIR::FacetAccessType>(expr_id)) {
-    facet_value_inst_id = facet_access_type->facet_value_inst_id;
-  } else if (context.types().Is<SemIR::FacetType>(type_of_expr_id)) {
-    facet_value_inst_id = expr_id;
-  }
+  CARBON_CHECK(context.types().IsFacetType(target_type_id));
+
+  // If the source type is/has a facet value (converted with `as type` or
+  // otherwise), then we can include its FacetType in the diagnostic to help
+  // explain what interfaces the source type implements.
+  expr_id = GetCanonicalFacetOrTypeValue(context, expr_id);
```

Comment:
Perhaps make a separate local variable here, so that it's clearer in uses that they're not using the argument?

```suggestion
  SemIR::InstId canonical_expr_id = GetCanonicalFacetOrTypeValue(context, expr_id);
```